### PR TITLE
Use debug-mode in CI testing

### PR DESCRIPTION
**Pull request by JoshuaLampert**
I've noticed (e.g. [here](https://github.com/trixi-framework/P4est.jl/pull/90#issuecomment-1763365340)) that `p4est` throws some errors in our tests, when debug-mode is enabled. I therefore suggest to enable `--enable-debug` in our CI testing as proposed in this PR. I have also adjusted our tests such that `p4est` doesn't throw the errors anymore. This is also related to https://github.com/trixi-framework/Trixi.jl/issues/1358 since there `p4est` also throws an error if debug-mode is enabled, although I'm not sure why.

### CODE REVIEW

File: test/test_nested_attributes_2d.jl [hunk 1]
```diff
@@ -18,7 +18,7 @@ end
 
 function refine_fn(p4est, which_tree, quadrant)
   quadrant_obj = unsafe_load(quadrant)
-  if quadrant_obj.x == 0 && quadrant_obj.y == 0 && quadrant_obj.level < 4
+  if quadrant_obj.x == 0 && quadrant_obj.y == 0 && quadrant_obj.level < 2
```
Comment (ranocha) [inline test/test_nested_attributes_2d.jl:21]: Why is this change necessary?

Comment (JoshuaLampert) [reply test/test_nested_attributes_2d.jl]: Otherwise, we get  [this](https://github.com/trixi-framework/P4est.jl/actions/runs/6523830565/job/17714856539?pr=90#step:11:643) error, i.e. [this assertion in p4est](https://github.com/cburstedde/p4est/blob/master/src/p4est_iterate.c#L2347) fails. I am not an expert of the internals of p4est to fully explain *why* this fails, however balancing the forest before iterating over it, fixes the problem (see [ae6e0d3](https://github.com/trixi-framework/P4est.jl/pull/98/commits/ae6e0d322c21932f901733d2382f1b7dff1336aa)) and is probably the proper fix. Iterating over a forest with neighboring trees that differ in more than level does not work (or something along those lines).

Comment (ranocha) [reply test/test_nested_attributes_2d.jl]: Thanks 👍
